Add `enum` support

alecgibson opened the pull request:
The `enum` is a popular aspect of TypeScript, and can be embedded in
interfaces as a concise, but descriptive, shorthand for literal string
unions:

```typescript
enum Colour {
  White = '000000',
  Black = 'ffffff'
}
```

This change adds an exported member `enum` to `io-ts`, based on
[this suggestion][1] by @noe132

It means that `enum`s can be reused directly in `io-ts`:

```typescript
const T = t.enum(Colour)
```

[1]: https://github.com/gcanti/io-ts/issues/216#issuecomment-471497998

alecgibson commented:
One place this does seem to fall down is that you lose the casting when extracting an interface. I can't quite figure out how to do that (possibly something to do with the `encode` function?). For example:

```typescript
enum Colour {
  White = '000000'
}

const Style = t.type({
  colour: t.enum(Colour)
})

interface IStyle extends t.TypeOf<typeof Style> {}

const style: IStyle = {
  colour: Colour.White
}

const colour = style.colour // colour has type "any" rather than "Colour"
```

I tried tweaking the signature for `enumType`:

```typescript
const enumType = <E, A = E[keyof E]>(e: E, name: string = 'Enum'): EnumType<A> => {
  const is = (u: unknown): u is A => Object.keys(e).some((k) => e[k as keyof E] === u)
  return new EnumType<A>(name, is, (u, c) => (is(u) ? success(u) : failure(u, c)), identity)
}
```

...which seems to only succeed in disallowing a `decode` to `string` (which I guess is correct?), but still doesn't correctly cast.

jessmorecroft commented:
This looks like it might work better, for string enums at least...

```
const enumType = <E>(
  e: { [key: string]: E },
  name: string = 'Enum',
): EnumType<E> => {
  const is = (u: unknown): u is E => Object.keys(e).some(k => e[k] === u);
  return new EnumType<E>(
    name,
    is,
    (u, c) => (is(u) ? t.success(u) : t.failure(u, c)),
    t.identity,
  );
};
```

saa18877433 commented:
Re: [gcanti/io-ts] Add `enum` support (#366)

2020-02-16 2:56 غرينتش-08:00, jessmorecroft <notifications@github.com>:
> This looks like it might work better, for string enums at least...
>
> ```
> const enumType = <E>(
>   e: { [key: string]: E },
>   name: string = 'Enum',
> ): EnumType<E> => {
>   const is = (u: unknown): u is E => Object.keys(e).some(k => e[k] === u);
>   return new EnumType<E>(
>     name,
>     is,
>     (u, c) => (is(u) ? t.success(u) : t.failure(u, c)),
>     t.identity,
>   );
> };
> ```
>
> --
> You are receiving this because you are subscribed to this thread.
> Reply to this email directly or view it on GitHub:
> https://github.com/gcanti/io-ts/pull/366#issuecomment-586692616

noe132 commented:
Hey, since number enums have remapping,
https://www.typescriptlang.org/docs/handbook/enums.html#reverse-mappings
I think it's essential to check if the key is the remapped one.
```typescript
enum E {
  A = 1
}

// compiles to
const E {
  A: 1,
  '1': 'A',
}
```
In this case, `1` is valid, but `'A'` is not.
https://www.typescriptlang.org/play?#code/FAUwHgDg9gTgLgAhAOwK4FsEFEEG9gIICCCAvAgIzAC+QA

I've updated my code here https://github.com/gcanti/io-ts/issues/216#issuecomment-471497998

abrkn commented:
Any update on this?

aroman commented:
+1 this would be a great addition, as currently I need to manually copy/paste my enum values into a `t.union` of `t.literal`s

alecgibson commented:
Okay I've updated this, and it now correctly casts!

<img width="432" alt="Screen Shot 2021-01-04 at 16 40 49" src="https://user-images.githubusercontent.com/12036746/103557822-a9cb9e00-4eab-11eb-9a76-b7c4f8a1c265.png">

Also updated to include @noe132 's update for checking for reverse mapping

Andrapyre commented:
Any update on this from @gcanti ?

halbich commented:
Hi, what is the current status of this feature? We'd love to have enum support :)

luisroel91 commented:
@gcanti How do you feel about merging this in?

martindines commented:
Is there anything stopping this feature from being merged?
